src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingResult.cs

```diff
@@ -1,47 +1,55 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+
 namespace Microsoft.AspNet.Mvc.ModelBinding
 {
     /// <summary>
     /// Contains the result of model binding.
     /// </summary>
-    public class ModelBindingResult
+    public struct ModelBindingResult : IEquatable<ModelBindingResult>
     {
-        /// <summary>
-        /// Creates a new <see cref="ModelBindingResult"/> indicating a fatal error.
-        /// </summary>
-        /// <param name="key">The key using which was used to attempt binding the model.</param>
-        public ModelBindingResult(string key)
+        public static readonly ModelBindingResult NoResult = new ModelBindingResult();
+
+        public static readonly Task<ModelBindingResult> NoResultAsync = Task.FromResult(NoResult);
+
+        public static ModelBindingResult Failed(string key)
         {
-            Key = key;
+            return new ModelBindingResult(key, model: null, isModelSet: false, validationNode: null);
         }
 
-        /// <summary>
-        /// Creates a new <see cref="ModelBindingResult"/>.
-        /// </summary>
-        /// <param name="model">The model which was created by the <see cref="IModelBinder"/>.</param>
-        /// <param name="key">The key using which was used to attempt binding the model.</param>
-        /// <param name="isModelSet">A value that represents if the model has been set by the
-        /// <see cref="IModelBinder"/>.</param>
-        public ModelBindingResult(object model, string key, bool isModelSet)
-            : this(model, key, isModelSet, validationNode: null)
+        public static Task<ModelBindingResult> FailedAsync(string key)
         {
+            return Task.FromResult(Failed(key));
         }
 
-        /// <summary>
-        /// Creates a new <see cref="ModelBindingResult"/>.
-        /// </summary>
-        /// <param name="model">The model which was created by the <see cref="IModelBinder"/>.</param>
-        /// <param name="key">The key using which was used to attempt binding the model.</param>
-        /// <param name="isModelSet">A value that represents if the model has been set by the
-        /// <see cref="IModelBinder"/>.</param>
-        /// <param name="validationNode">A <see cref="ModelValidationNode"/> which captures the validation information.
-        /// </param>
-        public ModelBindingResult(object model, string key, bool isModelSet, ModelValidationNode validationNode)
+        public static ModelBindingResult Success(string key, object model)
```

Comment:
Should not be an encouraged pattern and would prefer if `validationNode: null` were explicit in the calling code. That is, nuke this overload and the shorter `SuccessAsync()`.